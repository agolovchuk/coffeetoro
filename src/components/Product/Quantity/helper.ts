
export function quantify(quantity: number, m: number, handler: (q: number) => void) {
  return (): void => {
    const q = Math.max(quantity + m, 1);
    if (quantity !== q) {
      handler(Math.max(quantity + m, 1));
    }
  }
}

export function maybyRemove(quantity: number, onRemove: () => void) {
  return () => {
    if (quantity === 0) {
      onRemove();
    }
  }
}