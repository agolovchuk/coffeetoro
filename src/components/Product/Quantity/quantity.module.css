.container {
  display: flex;
  align-items: center;
  margin: 0 15px;
}
.field {
  font-size: 24px;
  width: 70px;
  height: 100%;
  margin: 0 5px;
  text-align: center;
  font-weight: bold;
  color: #3B4E63;
  border: 0 none;
}
.btn {
  width: 40px;
  height: 40px;
  border-radius: 50%;
  position: relative;
  color: #fff;
  border: 0 none;
}
.btn::before, .btn::after {
  display: inline-block;
  background-color: #fff;
  border-radius: 2px;
  position: absolute;
  top: 0;
  bottom: 0;
  right: 0;
  left: 0;
  margin: auto;
}
.btn::before {
  content: "";
  width: 20px;
  height: 3px;
}
.minus {
  background-color: #5496EF;
}
.plus {
  background-color: #7BCBB0;
}
.plus::after {
  content: "";
  height: 20px;
  width: 3px;
}