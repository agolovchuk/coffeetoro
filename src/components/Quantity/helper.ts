
export function quantify(quantity: number, m: number, handler: (q: number) => void) {
  return () => handler(Math.max(quantity + m, 1));
}

export function valueNormalize(value: string): number {
  const n = parseInt(value, 10)
  return isNaN(n) ? 0 : n;
}

export function maybyRemove(quantity: number, onRemove: () => void) {
  return () => {
    if (quantity === 0) {
      onRemove();
    }
  }
}