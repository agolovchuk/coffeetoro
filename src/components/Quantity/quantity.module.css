.container {
  display: flex;
  align-items: center;
}

.field {
  font-size: 30px;
  width: 80px;
  height: 100%;
  margin: 0 5px;
  text-align: center;
}

.btn {
  width: 30px;
  height: 100%;
  position: relative;
}
.btn::before {
  font-size: 20px;
}

.minus::before {
  content: '-';
}
.plus::before {
  content: '+';
}

.remove {
  height: 100%;
  margin-left: 10px;
  background-color: red;
  width: 40px;
  border: 0 none;
}
.remove::before {
  content: 'x';
  font-size: 20px;
  color: #fff;
}