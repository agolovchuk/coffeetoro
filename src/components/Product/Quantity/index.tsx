import * as React from 'react';
import cx from 'classnames';
import { quantify, valueNormalize } from './helper';
import styles from './quantity.module.css';

interface Props {
  quantity: number;
  onChange: (q: number) => void;
  onRemove: () => void;
}

function Quantity({ quantity, onChange, onRemove }: Props) {
  const clickHandler = (m: number) => quantify(quantity, m, onChange);
  const changeHandler = (e: React.ChangeEvent<HTMLInputElement>) => onChange(valueNormalize(e.currentTarget.value))
  return (
    <div
      className={styles.container}
    >
      <button
        type="button"
        onClick={clickHandler(-1)}
        className={cx(styles.btn, styles.minus)}
      />
      <input
        type="text"
        pattern="\b"
        value={quantity}
        onChange={changeHandler}
        className={styles.field}
      />
      <button
        type="button"
        onClick={clickHandler(1)}
        className={cx(styles.btn, styles.plus)}
      />
    </div>
  )
}

export default Quantity;