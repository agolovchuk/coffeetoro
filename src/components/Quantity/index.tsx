import * as React from 'react';
import cx from 'classnames';
import { Modal, Popup, Confirmation } from 'components/Popup';
import { quantify, valueNormalize, maybyRemove } from './helper';
import styles from './quantity.module.css';

interface Props {
  quantity: number;
  onChange: (q: number) => void;
  onRemove: () => void;
}

function Quantity({ quantity, onChange, onRemove }: Props) {
  const clickHandler = (m: number) => quantify(quantity, m, onChange);
  const changeHandler = (e: React.ChangeEvent<HTMLInputElement>) => onChange(valueNormalize(e.currentTarget.value))
  const [confirm, setConfirm] = React.useState(false);
  return (
    <div
      className={styles.container}
    >
      <button
        type="button"
        onClick={clickHandler(-1)}
        className={cx(styles.btn, styles.minus)}
      />
      <input
        type="text"
        pattern="\b"
        value={quantity}
        onChange={changeHandler}
        className={styles.field}
      />
      <button
        type="button"
        onClick={clickHandler(1)}
        className={cx(styles.btn, styles.plus)}
      />
      <button
        type="button"
        className={styles.remove}
        onClick={() => setConfirm(true)}
      />
      {
        confirm ? (
          <Modal>
            <Popup onCancel={() => null}>
              <Confirmation
                title="Confirm remove"
                onCancel={() => setConfirm(false)}
                onConfirm={() => onRemove()}
              >
                <p>Realy?</p>
              </Confirmation>
            </Popup>
          </Modal>
        ) : null
      }
    </div>
  )
}

export default Quantity;