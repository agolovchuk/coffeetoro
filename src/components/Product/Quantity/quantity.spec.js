import * as React from 'react';
import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import sinon from 'sinon';

import Quantity from './index';

Enzyme.configure({ adapter: new Adapter() });

describe('Quantity', () => {
  let onChange;
  const quantity = 2;
  beforeEach(() => {
    onChange = sinon.spy();
  });

  it('Should be rendered correct', () => {
    const element = shallow(
      <Quantity quantity={quantity} onChange={onChange} />
    )
    expect(toJson(element)).toMatchSnapshot();
  });

  it('Should be have an active element', () => {
    const element = shallow(
      <Quantity quantity={quantity} onChange={onChange} />
    )
    expect(element.find('button.minus')).toHaveLength(1);
    expect(element.find('button.plus')).toHaveLength(1);
    expect(element.find('input')).toHaveLength(1);
    expect(element.find('input').props().value).toEqual(quantity);
  })

  it('Should called when clicked Plus', () => {
    const element = shallow(
      <Quantity quantity={quantity} onChange={onChange} />
    )
    const plus = element.find('button.plus');
    plus.simulate('click');
    expect(onChange.called).toBeTruthy();
    expect(onChange.args[0][0]).toEqual(quantity + 1);
  });

  it('Should called when click Minus', () => {
    const element = shallow(
      <Quantity quantity={quantity} onChange={onChange} />
    )
    const minus = element.find('button.minus');
    minus.simulate('click');
    expect(onChange.called).toBeTruthy();
    expect(onChange.args[0][0]).toEqual(quantity - 1);
  });

  it('Should be send correct value when change input', () => {
    const element = shallow(
      <Quantity quantity={quantity} onChange={onChange} />
    )
    const input = element.find('input');
    input.simulate('change', { target: { value: '6' } });
    expect(onChange.args[0][0]).toEqual(6);
    input.simulate('change', { target: { value: '' } });
    expect(onChange.args[0][0]).toEqual(6);
    input.simulate('change', { target: { value: 'f' } });
    expect(onChange.args[0][0]).toEqual(6);
    expect(onChange.calledOnce).toBeTruthy();
    input.simulate('change', { target: { value: '2' } });
    expect(onChange.calledOnce).not.toBeTruthy();
  });

});