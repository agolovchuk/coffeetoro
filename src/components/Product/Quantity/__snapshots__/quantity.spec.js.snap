// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quantity Should be rendered correct 1`] = `
<div
  className="container"
>
  <button
    className="btn minus"
    onClick={[Function]}
    type="button"
  />
  <input
    className="field"
    inputMode="numeric"
    onChange={[Function]}
    type="text"
    value={2}
  />
  <button
    className="btn plus"
    onClick={[Function]}
    type="button"
  />
</div>
`;
